backend-api/src/functions/adapters/dynamoDbAdapter.ts

```diff
@@ -0,0 +1,195 @@
+import {
+  AttributeValue,
+  ConditionalCheckFailedException,
+  DynamoDBClient,
+  PutItemCommand,
+  PutItemCommandInput,
+  QueryCommand,
+  QueryCommandInput,
+  QueryCommandOutput,
+} from "@aws-sdk/client-dynamodb";
+import { errorResult, Result, successResult } from "../utils/result";
+import {
+  CreateSessionAttributes,
+  SessionDetails,
+} from "../services/sessionService/sessionService";
+import { NodeHttpHandler } from "@smithy/node-http-handler";
+
+type DynamoDbRecord = Record<string, AttributeValue>;
+
+export class DynamoDbAdapter {
+  private readonly tableName: string;
+  private readonly dynamoDbClient = new DynamoDBClient({
+    region: process.env.REGION,
+    maxAttempts: 2,
+    requestHandler: new NodeHttpHandler({
+      connectionTimeout: 29000,
+      requestTimeout: 29000,
+    }),
+  });
+
+  constructor(tableName: string) {
+    this.tableName = tableName;
+  }
+
+  async readSessionId(
+    subjectIdentifier: string,
+  ): Promise<Result<string | null>> {
+    const currentTimeInSeconds = Math.floor(Date.now() / 1000);
+    const input: QueryCommandInput = {
+      TableName: this.tableName,
+      IndexName: "subjectIdentifier-timeToLive-index",
+      KeyConditionExpression:
+        "subjectIdentifier = :subjectIdentifier and :currentTimeInSeconds < timeToLive",
+      FilterExpression: "sessionState = :sessionState",
+      ExpressionAttributeValues: {
+        ":subjectIdentifier": { S: subjectIdentifier },
+        ":sessionState": { S: "ASYNC_AUTH_SESSION_CREATED" },
+        ":currentTimeInSeconds": {
+          N: currentTimeInSeconds.toString(),
+        },
+      },
+      ProjectionExpression: "sessionId",
+      Limit: 1,
+      ScanIndexForward: false,
+    };
+
+    const queryResponse = await this.query(input);
+    if (queryResponse.isError || queryResponse.value === null) {
+      return queryResponse;
+    }
+
+    const record: DynamoDbRecord = queryResponse.value;
+    const sessionId = record.sessionId?.S;
+    if (!sessionId) {
+      return errorResult({
+        errorMessage: "Session is malformed",
+        errorCategory: "SERVER_ERROR",
+      });
+    }
+
+    return successResult(sessionId);
+  }
+
+  async readSessionDetails(
+    subjectIdentifier: string,
+  ): Promise<Result<SessionDetails | null>> {
+    const currentTimeInSeconds = Math.floor(Date.now() / 1000);
```

Comment:
We've done this twice in this function - how about a small function within this file?

Reply:
Done...